[WIP] Single database multitenancy

leela opened the pull request:
**List of changes**
1. Row level Security is added
2. site creation flow is changed
3. new tenant CLI is added 
4. Added new flags and new hooks

In a current setup,  new site flow takes care creating DB tables and inserting records into DB (ex: roles, users etc) of an app. The records those are inserted are supposed to be part of tenant flow in case of multi-tenancy.

Incase of multi-tenancy, Tables are divided into site specific tables and tenant specific tables. Tenant specific tables will have tenant_id in those and RLS takes care of cross-tenant data access issues.

Every app installation also involves two steps, one is site specific setup and the other is tenant specific setup. When a new site is created, only site specific setup of every app will be executed. Tenant specific setup is executed when new tenant added to the system.

In a new flow, while creating a new site we create schema from doctypes and changes relates to schema(adding doctype field information in DocField table etc.). All record insertions(Ex: creating roles or users etc) takes part in new tenant flow. 

P.S: 
1. I am not so good at naming stuff. Please feel free to pass your comments if you feel naming is inappropriate.
2. And you see tags like `FIXME` & `TODO` in comments. Those are reminders for me to fix. I will take care of those.

leela committed:
feat: Tenant table added

As part of singledb-multitenancy, New internal tenant table is added.

leela committed:
feat(singledb-multitenancy): tenant_id col added to postgres tables

TenantID is added to all tenant specific tables in DB.

leela committed:
feat(singledb): split site&tenant creation logics

In current multidb-multitenant architecture, site and tenant creation
logic are interwined. Splitting that for singledb-multitenancy
architecture.

leela committed:
feat(singledb): add composit unique constraints

Incase of tenant specific tables, `tenant_id` has to be part of all
unique constraints. Made composite unique constraints incase of tenant
specific tables.

leela committed:
feat(singledb): tenant_id col in few internal tables

TenantID is added to all tenant specific tables in DB.

leela committed:
chore: Indentation settings added for sql files

leela committed:
feat(singledb): Tenant db module and CLI

Added placeholder of Tenant CLI and setting up tenant in the db session.

leela committed:
feat(singledb): site creation flow changes

As part of multi-tenancy support, we split current site flow into site
flow and new tenant flow.

In new site flow, we create DB tables and meta details about those
tables (Record creation is moved into tenant flow).

leela committed:
feat(singledb): Conf runtime parameters to the current db conn session

leela committed:
feat(singledb): Force RLS on internal tables post site creation

leela committed:
feat(singledb): Add tenant CLI added

leela committed:
feat(SingleDB): refactored tenant and new site CLI

In a new flow, app instalation invaloves two things.
1. Install site specifics from the app
2. Instal tenant specifics from app

We do install site specifics from the app when new site is created and
tenant specifics when ned tenant is added to the system.

Added new hook named `after_tenant_specifics_install`.

rmehta commented on frappe/core/doctype/doctype/doctype.py:
```diff
@@ -1122,19 +1127,30 @@ def check_if_importable(d):
+			raise
+
+def make_roles(doc, perm_fieldname="permissions"):
+	"""Make `Role` records if already not made. Called while installing."""
```
indent seems broken?

rmehta commented on frappe/installer.py:
```diff
@@ -115,78 +106,132 @@ def install_db(root_login="root", root_password=None, db_name=None, source_sql=N
 
 	frappe.flags.in_install_db = False
 
+def plugin_site_specifics_from_app(app: str, verbose=False) -> None:
```
why rename this function? `plugin_site_specifics_from_app` seems very cryptic

rmehta replied:
`install_app_for_tenant` ?

rmehta replied:
common fixtures can be added via `install_app`
tentant specific fixtures can be added via `install_tenant`

by default, we will install one tenant

leela replied:
you mean `install_app_for_site` here?
 
I thought of it but I was unsure about having `install_app_for_tenant` and `install_app_for_site` names as it din't feel correct to me. will fix that!

rmehta commented on frappe/database/postgres/schema.py:
```diff
@@ -26,10 +38,64 @@ def create(self):
+		frappe.db.sql(sql.strip(',\n'))
 		frappe.db.commit()
 
+	def create_tenant_specific_table(self):
```
The 2 methods can be

`create_table`
`create_table_multitenant`

rmehta commented on frappe/installer.py:
```diff
@@ -115,78 +106,132 @@ def install_db(root_login="root", root_password=None, db_name=None, source_sql=N
+	frappe.flags.in_site_setup = False
+	frappe.flags.in_install = False
+
+def plugin_tenant_specifics_from_app(app: str, tenant: Tenant, verbose: bool=False,
```
`install_tentant`

leela committed:
feat(SingleDB): Initialize singles while adding tenant

leela committed:
feat(SingleDB): Create Guest tenant while site creation

Current login flow is heavily depends on User and other tables, but in
multi-tenancy unless we know the tenant we can not use tenant tables.
As a patch, adding a Guest tenant. All non-loggedin requests use
Guest tenant in the flow.

leela committed:
feat(SingleDB): Naming functions appropriately

leela committed:
feat(SingleDB): Include tenantID in redis key

leela committed:
feat(singleDB): Added tenant as arg to some of CLIs

tenant information is needed for some of the CLI's to function normally.
CLI's like add-system-manager, disable-user, set_last_active_for_user
etc can't ben used without tenant. Added tenant argument for those
CLI's.

leela committed:
fix: do not set user while initializing tenant

leela committed:
fix: Sider issues fixed

leela committed:
fix: mariadb site login issue bcs of session

Login is failing in mariadb because of missing tenant_id in session table.
tenant_id is added to mariadb session schema.

leela committed:
refactor: renaming functions in schema module

leela commented:
Halting for now. Will re-open when time comes.
